pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,172 @@
+            # \lambda_{c \rightarrow \s} = 1/|S(c)| * max_{x_{c-s}}{\theta_{ij}}
+            self.message_from_cluster = {}
+            for i in self.intersection_sets_for_cluster_c:
+                other_variables = list(set(set_of_variables_for_cluster_c)-i)
+                phi = copy.deepcopy(cluster_potential)
+
+                # phi = max_{x_{c-s}}{\theta_{ij}}
+                phi.maximize(other_variables)
+                self.message_from_cluster[i] = (1 / len(self.intersection_sets_for_cluster_c)) * phi
+
+    def _update_message(self, sending_cluster):
+
+        """
+        This is the message-update method.
+
+        Parameters
+        ----------
+        sending_cluster: The resulting messages are \lambda_{c \rightarrow s} from the given
+            cluster 'c' to all of its intersection_sets 's'.
+            Here 's' are the elements of intersection_sets_for_cluster_c.
+
+        Reference
+        ---------
+        Fixing Max-Product: Convergent Message-Passing Algorithms for MAP LP Relaxations
+        by Amir Globerson and Tommi Jaakkola.
+        Section 6, Page: 5; Beyond pairwise potentials: Generalized MPLP
+        """
+        updated_results = []
+        # The new updates will take place for the intersection_sets of this cluster.
+        for i in sending_cluster.intersection_sets_for_cluster_c:
+
+            # Now we have to construct the crux of the equation which is used to update the messages there.
+
+            # The terms which we need are:
+            # 1. Message not originating from the current sending_cluster but going into the current intersect i
+            message_not_from_cluster = []
+
+            # 2. Summation of messages not originating from the current sending_cluster and going into other
+            #    intersects than i of this cluster only.
+            sum_other_intersects = []
+
+            cluster_list_except_current = [j for j in self.cluster_set if j != sending_cluster]
+            other_intersects = [j for j in sending_cluster.intersection_sets_for_cluster_c if j != i]
+
+            for cluster in cluster_list_except_current:
+                for intersection_set in cluster.intersection_sets_for_cluster_c:
+                    if i == intersection_set:
+                        message_not_from_cluster.append(cluster.message_from_cluster[intersection_set])
+
+                for other_intersect in other_intersects:
+                        for intersection_set in cluster.intersection_sets_for_cluster_c:
+                            if other_intersect == intersection_set:
+                                sum_other_intersects.append(cluster.message_from_cluster[intersection_set])
+
+            # lambda_1 = \lambda_{s}^{-c}
+            lambda_1 = copy.deepcopy(sending_cluster.message_from_cluster[i])
+            lambda_1.values = np.zeros(np.prod(lambda_1.cardinality))
+            # lambda_2 = \sum_{}{\lambda_{s'}^{-c}}
+            lambda_2 = copy.deepcopy(lambda_1)
```

Comment:
@kislayabhi Lets discuss on how to create a method in Factor class for copying. I don't using ```copy.deepcopy``` method so many times.